datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -415,6 +441,74 @@ public void onPartitionMovement(Long notifyTimestamp) {
     _eventQueue.put(CoordinatorEvent.createPartitionMovementEvent(notifyTimestamp));
     _log.info("Coordinator::onPartitionMovement completed successfully");
   }
+
+  /**
+   * {@inheritDoc}
+   * Stop all the tasks and wait for new session to connect.
+   */
+  @Override
+  public void onSessionExpired() {
+    if (_shutdown) {
+      return;
+    }
+    stopEventThread();
+
+    _leaderDatastreamAddOrDeleteEventScheduled.set(false);
+    if (_leaderDatastreamAddOrDeleteEventScheduledFuture != null) {
+      _leaderDatastreamAddOrDeleteEventScheduledFuture.cancel(true);
+      _leaderDatastreamAddOrDeleteEventScheduledFuture = null;
+    }
+
+    _leaderDoAssignmentScheduled.set(false);
+    if (_leaderDoAssignmentScheduledFuture != null) {
+      _leaderDoAssignmentScheduledFuture.cancel(true);
+      _leaderDoAssignmentScheduledFuture = null;
+    }
+
+    _eventQueue.clear();
+
+    // Stopping all the connectors so that they stop producing.
+    List<Future<Boolean>> assignmentChangeFutures = _connectors.keySet().stream()
+        .map(connectorType -> {
+          _assignmentChangeThreadPool.get(connectorType).shutdownNow();
+          _assignmentChangeThreadPool.put(connectorType, Executors.newSingleThreadExecutor());
+          return dispatchAssignmentChangeIfNeeded(connectorType, new ArrayList<>(), false, false);
+        })
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+
+    // Wait till all the futures are complete or timeout.
+    ExecutorService threadPoolExecutor = Executors.newFixedThreadPool(1);
+    threadPoolExecutor.submit(() -> {
+      Instant start = Instant.now();
+      try {
+        getAssignmentsFuture(assignmentChangeFutures, start);
+      } catch (Exception e) {
+        // if it's timeout then we will retry
```

Comment:
Where is the retry logic? I don't see any retries happening.
In general, what should the protocol be for TimeoutException? And other exceptions like interrupt? Can this get interrupted?

Basically, we need the tasks to stop, what if they don't? E.g. assignment taking too long so we cancel the futures, or our thread gets interrupted. The debounce timer assumption for the persistent lock changes is that the tasks will be brought down in 30 seconds.